transactions/src/main/java/org/ehcache/transactions/xa/XAStore.java

```diff
@@ -0,0 +1,897 @@
+          }
+        };
+        return;
+      }
+
+      while (underlyingIterator.hasNext()) {
+        final Cache.Entry<K, ValueHolder<SoftLock<V>>> next = underlyingIterator.next();
+
+        if (!transactionContextFactory.isTouched(transactionId, next.getKey())) {
+          ValueHolder<SoftLock<V>> valueHolder = next.getValue();
+          SoftLock<V> softLock = valueHolder.value();
+          final XAValueHolder<V> xaValueHolder;
+          if (softLock.getTransactionId() == transactionId) {
+            xaValueHolder = new XAValueHolder<V>(valueHolder, softLock.getNewValueHolder().value());
+          } else if (isInDoubt(softLock)) {
+            continue;
+          } else {
+            xaValueHolder = new XAValueHolder<V>(valueHolder, softLock.getOldValue());
+          }
+          this.next = new Cache.Entry<K, ValueHolder<V>>() {
+            @Override
+            public K getKey() {
+              return next.getKey();
+            }
+
+            @Override
+            public ValueHolder<V> getValue() {
+              return xaValueHolder;
+            }
+
+            @Override
+            public long getCreationTime(TimeUnit unit) {
+              return next.getCreationTime(unit);
+            }
+
+            @Override
+            public long getLastAccessTime(TimeUnit unit) {
+              return next.getLastAccessTime(unit);
+            }
+
+            @Override
+            public float getHitRate(TimeUnit unit) {
+              return next.getHitRate(unit);
+            }
+          };
+          break;
+        }
+      }
+    }
+
+    @Override
+    public boolean hasNext() throws CacheAccessException {
+      try {
+        if (!getCurrentContext().getTransactionId().equals(transactionId)) {
+          throw new IllegalStateException("Iterator has been created in another transaction, it can only be used in the transaction it has been created in.");
+        }
+      } catch (XACacheAccessException xacae) {
+        throw xacae.getCause();
+      }
+      return next != null;
+    }
+
+    @Override
+    public Cache.Entry<K, ValueHolder<V>> next() throws CacheAccessException {
+      try {
+        if (!getCurrentContext().getTransactionId().equals(transactionId)) {
+          throw new IllegalStateException("Iterator has been created in another transaction, it can only be used in the transaction it has been created in.");
+        }
+      } catch (XACacheAccessException xacae) {
+        throw xacae.getCause();
+      }
+      if (next == null) {
+        throw new NoSuchElementException();
+      }
+      Cache.Entry<K, ValueHolder<V>> rc = next;
+      advance();
+      return rc;
+    }
+  }
+
+  @Override
+  public ValueHolder<V> compute(K key, BiFunction<? super K, ? super V, ? extends V> mappingFunction, NullaryFunction<Boolean> replaceEqual) throws CacheAccessException {
+    checkKey(key);
+    XATransactionContext<K, V> currentContext = getCurrentContext();
+    if (currentContext.touched(key)) {
```

Comment:
Inside the `currentContext.touched(key)` the code is the same. Maybe you need a condition before invoking the passed in function in `computeIfPresent`?